rfcs/013 - Datastore Revisited RFC.md

```diff
@@ -0,0 +1,142 @@
+
+There are multiple ways to resolve this such that the actual datastore opcodes are left alone while the compiler guarantees that their usage is safe. For the sake of unblocking datastore usage as fast as possible, we're going to go through a two-step process to get it working, if a bit weird, and then using a new language feature to implement it cleanly. This does mean that the API will be changed twice, but since we are so far away from 1.0, this is considered fine for the adventurous who want to use this standard library that isn't documented publicly, yet.
+
+### Part 1: Minimally-invasive with restored type introspection
+
+We recently removed the ability to get the type of a variable as a string with the `type variableName` syntax, but if we restore that, it should be possible to force type safety by combining the type string with either the namespace or the key string, such that trying to get the same key name with a default value of a different type will find that it doesn't exist and return the provided default.
+
+This will provide zero actual API call changes except to the `has` and `del` functions, which would need to be given an example of the type of data to delete in order to find it. This has the smallest change requirement, though the API is a bit weird.
+
+### Part 2: Generic Functions (in addition to the interface functions that currently exist)
+
+Functions gain the ability to have one or more generic types, eg `fn foo<A>(arg1: int64, arg2: string): A {` such that they can be called with some type that is passed in during invocation but is not passed in as a variable, just the type, so `foo<bool>(1, 'bar')` would do something internally with the type.
+
+In this case, the `namespace` function would take not only the name parameter, but the type info for the kind of variable to store in it, like `namespace<Array<int64>>('variousDigitsOfPi')`, then this type would internally be applied to the namespace using a syntax similar to `type variableName` (perhaps as simple as `type GenericName`), so it can be stored in the datastore opcodes. This would remove the need to provide example values or type information to the `has` and `del` functions since it would already exist on the provided namespace.
+
+It'll make the language much more complex, but the utility should surpass the complexity. This approach could be turned into the foundation for interface functions, such that `fn someFn<A as Orderable>(a: A, b: A): A` could be equivalent to `fn someFn(a: Orderable, b: Orderable): Orderable` with the latter becoming syntactic sugar. This would also allow us to remove duplicated interfaces like `anythingElse` because `fn map<A as Any, B as Any>(arr: Array<A>, mapper: fn(A): B): Array<B>` makes it clear that while `A` and `B` match the same interface, they are not the same actual type.
```

Comment:
I think it should be noted that the desugaring of the second example of `someFn` wouldn't desugar into the first example. ie `someFn(a: Orderable, b: Orderable): Orderable` *shouldn't* desugar into `fn someFn<A as Orderable>(a: A, b: A): A`, but rather `fn someFn<A as Orderable, B as Orderable, R as Orderable>(a: A, b: B): R`

Reply:
Actually, you're wrong there, it *will* be equivalent to what I wrote in the RFC.

Reply:
I was thinking there might be a case where the desugaring as written might not be what's expected, but I can't think of one. Makes sense